### PR TITLE
[New] Add hyperHtml w/ npm auto-update via single package.json

### LINKED ISSUE #12546: [Request] Add hyperHTML
**Library name:** hyperHTML

**Library description:** 
A Fast & Light Virtual DOM Alternative available for NodeJS and NativeScript too.

**Git repository url:** 
https://github.com/WebReflection/hyperHTML 

**npm package name or url**: 
https://www.npmjs.com/package/hyperhtml

**License (List them all if it's multiple):** 
ISC License

**Official homepage:**
https://viperhtml.js.org/

**Wanna say something? Leave message here:** 

=====================

Notes from cdnjs maintainer(please remove this section after you read it):
1. Please read the [README.md](https://github.com/cdnjs/cdnjs#cdnjs-library-repository) and [CONTRIBUTING.md](https://github.com/cdnjs/cdnjs/blob/master/CONTRIBUTING.md) document first.
2. For the new library request issue, please make sure it's not a personal project, we have a basic requirement for the popularity, like 200 stars on GitHub or 800 downloads/month on npm registry.

We encourage you to add a library via sending pull request,
it'll be faster than just opening a request issue,
since there are tons of issues, please wait with patience,
and please don't forget to read the guidelines for contributing, thanks!!

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/55475902-request-add-hyperhtml?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by James-N-M**
Pull request for issue: #12546

Checklist for **Pull request** or **lib adding request issue** follows the conventions.

Note that if you are using a distribution purpose repository/package, please also provide the url and other related info like popularity of the source code repo/package.

# Profile of the lib
 * Git repository (required): https://github.com/WebReflection/hyperHTML
 * Official website (optional, not the repository): https://viperhtml.js.org/hyper.html
 * NPM package url (optional): https://www.npmjs.com/package/hyperhtml
 * License and its reference: ISC License
 * GitHub / Bitbucket popularity (required):
   - Count of watchers: 78
   - Count of stars: 1,476
   - Count of forks: 51
 * NPM download stats (optional):
   - Downloads in the last day: 411
   - Downloads in the last week: 3101
   - Downloads in the last month: 10484

# Essential checklist
 * [ ] I'm the author of this library
   * [ ] I would like to add link to the page of this library on CDNJS on website / readme
 * [x] This lib was not found on cdnjs repo
 * [x] Doesn't already exist / has duplicate issue and PR
 * [x] The lib has notable popularity
   * [x] More than 200 [Stars / Watchers / Forks] on [GitHub / Bitbucket]
   * [x] More than 800 downloads stats per month on npm registry
 * [x] Project has public repository on famous online hosting platform (or been hosted on npm)

# Auto-update checklist
 * [x] Has valid tags for each versions (for git auto-update)
 * [x] Auto-update setup
 * [x] Auto-update target/source is valid.
 * [x] Auto-update filemap is correct.

# Git commit checklist
 * [x] The first line of commit message is less then 50 chars; clean, clear and easy to understand.
 * [x] The parent of the commit(s) in the PR is not older than 3 days.
 * [x] Pull request is sent from a non-master branch with a meaningful name.
 * [x] Separate unrelated changes into different commits.
 * [x] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.
 * [x] Close corresponding issue in commit message
 * [x] Mention related issue(s), people in commit message, comment.

### CODE REVIEW

File: ajax/libs/hyperHtml/package.json [hunk 1]
```diff
@@ -1,33 +1,33 @@
...
-    "npmName": "hyperhtml",
-    "npmFileMap": [
-     {
-       "basePath": "dist",
-       "files": [
-         "**/*"
-       ]
-     }
+  "name": "hyperHtml",
+  "filename": "min.min.js",
```
Comment (WebReflection) [inline ajax/libs/hyperHtml/package.json]: oh gosh ... really? it's just min.js in the repo :smile:

Comment (James-N-M) [reply ajax/libs/hyperHtml/package.json]: @WebReflection lol I kinda just guessed what they meant by "main file" and the convention said to use filename.min.js so.... ya I dunno if this is right but you'd probs know best lol ! so I change this to min.js ???

**Comment by James-N-M** [comment]
@PeterDaveHello Hey Peter, I cant seem to get the format write even after using a Linter with two spaces, i think it might have something to do with the github link. Let me know ! :D

File: ajax/libs/hyperHtml/package.json (hunk 1 above)
Comment (James-N-M) [reply ajax/libs/hyperHtml/package.json]: thanks @WebReflection I'm still confused a bit so do I just take the filename out ??

Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: I've no idea what this bot expect, sorry :cry:

Comment (James-N-M) [reply ajax/libs/hyperHtml/package.json]: haha its all good ! :) 🏆  Ill wait for one of the reviews to check it out, I guess somethings wrong with my formatting ?? tabs vs spaces 👾  things lol I used a "Linter" but that still doesn't seem to do the trick and I have no idea how to access the cdnjs tool they mention.

**Comment by PeterDaveHello** [comment]
@James-N-M also the "repository" section in package.json, "type" & "url" has wrong indentation

**Comment by James-N-M** [comment]
@PeterDaveHello I tried again, if it doesnt pass how can i use the cdn js tool to do it for me, lol just having a little trouble.

**Comment by PeterDaveHello** [comment]
@James-N-M the tools only run locally, I'm not sure if you've cloned or your want to clone the repository locally, I believe that you can refer to the result in CI to help confirm the corrent style of json.

File: ajax/libs/hyperHtml/package.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "hyperHtml",
+  "filename": "min.min.js",
```
Comment (extend1994) [inline ajax/libs/hyperHtml/package.json]: We can just use `min.js` because it's already minified :slightly_smiling_face:

File: ajax/libs/hyperHtml/package.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    "alternative"
+  ],
+  "npmName": "hyperhtml",
+  "npmFileMap": [
+    {
+      "basePath": "dist",
+      "files": [
+        "**/*"
+      ]
+    }
```
Comment (extend1994) [inline ajax/libs/hyperHtml/package.json]: Suppose the file for a browser is only `min.js`. It's under the root directory, so the pattern here would be
```js
   {
      "basePath": "",
      "files": [
        "min*.+(css|js|map)"
      ]
    }

```

@WebReflection Could you tell me if `hyperhtml.js` in the old versions is the unminified version of `min.js`? If the answer is yes, we hope to include it in our auto-update config too.

File: ajax/libs/hyperHtml/package.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "hyperHtml",
+  "filename": "min.min.js",
+  "description": "A Fast & Light Virtual DOM Alternative",
+  "license": "ISC",
+  "author": "Andrea Giammarchi",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/WebReflection/hyperhtml.git"
+  },
```
Comment (extend1994) [inline ajax/libs/hyperHtml/package.json]: Let's also add homepage for it! Thanks!
```js
"homepage": "https://viperhtml.js.org/hyper.html",
```

File: ajax/libs/hyperHtml/package.json (hunk 2 above)
Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: see https://github.com/cdnjs/cdnjs/pull/12547#pullrequestreview-99160494

File: ajax/libs/hyperHtml/package.json (hunk 3 above)
Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: @extend1994 from https://unpkg.com/hyperhtml/ , the following are the files that makes sense to have in a CDN:

  * `/esm/**`
  * `/index.js`
  * `/min.js`
  * `/no.js`
  * `/umd.js`

Did I answer your question?

Comment (extend1994) [reply ajax/libs/hyperHtml/package.json]: @WebReflection Yes :+1: Just still have one question: `esm` files seem unable to be used in a browser? Like the way `<script src="*.js">`. Not sure if my knowledge is correct :sweat:

Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: Welcome to the future of JavaScript modules ;-)

```html
<!doctype html>
<script type=module>
import {bind} from 'https://unpkg.com/hyperhtml/esm/index.js';

window.onload = () => {
  bind(document.body)`Hello World`;
};
</script>
```

TL;DR Yes, `/esm/**/*` folder is usable right away to import ESM (ECMAScript Modules)

Comment (extend1994) [reply ajax/libs/hyperHtml/package.json]: @WebReflection Thanks a lot for the explanation! Good to know it :smile: However, I can't be sure that cdnjs can host this kind of files. To my knowledge, cdnjs hasn't hosted them yet.
@PeterDaveHello Could you please help check this? Thanks!

Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: FWIW, bundlers make the usage of these files via CDN somehow useless, so that having the following might be enough for the time being:

  * /index.js
  * /min.js
  * /no.js
  * /umd.js

Comment (extend1994) [reply ajax/libs/hyperHtml/package.json]: @PeterDaveHello May I know your opinion about the files of ECMAScript modules which is mentioned above?

Comment (MattIPv4) [inline ajax/libs/hyperHtml/package.json]:
```suggestion
    {
      "basePath": "",
      "files": [
        "*.js"
      ]
    }
```

This appears to no longer be valid for more recent versions -- it would seem the required files are now at the root of the package:

```
tree node_modules/hyperhtml                                                                                                                                                            ✔  9023  13:15:19 
node_modules/hyperhtml
├── CHANGELOG.md
├── CONTRIBUTING.md
├── HELPERS.md
├── LICENSE.txt
├── README.md
├── cjs
│   ├── classes
│   │   └── Component.js
│   ├── hyper
│   │   ├── render.js
│   │   └── wire.js
│   ├── index.js
│   ├── objects
│   │   ├── Intent.js
│   │   └── Updates.js
│   ├── package.json
│   └── shared
│       ├── constants.js
│       └── utils.js
├── esm
│   ├── classes
│   │   └── Component.js
│   ├── hyper
│   │   ├── render.js
│   │   └── wire.js
│   ├── index.d.ts
│   ├── index.js
│   ├── objects
│   │   ├── Intent.js
│   │   └── Updates.js
│   └── shared
│       └── constants.js
├── esm.js
├── index.d.ts
├── index.js
├── logo.txt
├── min.js
├── no.js
├── package.json
├── rollup.ie.js
├── umd.d.ts
└── umd.js
```

Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: The only source of truth in any of my modules is the `esm/` folder. All scripts are generated from that folder, but the file that should be served via CDNs is the one specified in the `"unpkg": "min.js"` field. The `min.js` is already babelized and minified, and that's indeed what https://unpkg.com/hyperhtml serves to have a global `hyperHTML` reference.

To use the ESM variant, https://unpkg.com/hyperhtml?module points at the ESM entry point, which is the `esm/index.js`, but it also resolves automatically all imported sub-modules.

I hope this helps.

Comment (MattIPv4) [reply ajax/libs/hyperHtml/package.json]: @WebReflection - So we should bring in `min.js` and then all the files contained within `esm`?

Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: @MattIPv4 the `min.js` is the bundle already minified, the `index.js` is the same bundle pre-minification. `esm/index.js` is for native ESM modules, but it requires some resolution of itse dependencies, so if CDNJS cannot do such resolution, being static, `min.js` and `index.js` is all it should have, excluding everything else.

Comment (WebReflection) [reply ajax/libs/hyperHtml/package.json]: P.S. worth saying *maybe* `umd.js` could be included too, for weird/edge cases.

Comment (MattIPv4) [reply ajax/libs/hyperHtml/package.json]: We support ts, so I'll include the typings as well.

**Comment by MattIPv4** [comment]
Replaced with #13742
